editor/src/components/navigator/navigator-utils.ts

```diff
@@ -0,0 +1,204 @@
+import {
+  ElementInstanceMetadataMap,
+  isJSXConditionalExpression,
+  isJSXFragment,
+  JSXConditionalExpression,
+} from '../../core/shared/element-template'
+import { MetadataUtils } from '../../core/model/element-metadata-utils'
+import { foldEither, isRight } from '../../core/shared/either'
+import {
+  conditionalClauseNavigatorEntry,
+  isConditionalClauseNavigatorEntry,
+  NavigatorEntry,
+  navigatorEntryToKey,
+  regularNavigatorEntry,
+  syntheticNavigatorEntry,
+} from '../editor/store/editor-state'
+import {
+  buildTree,
+  ElementPathTree,
+  ElementPathTreeRoot,
+  getSubTree,
+  reorderTree,
+} from '../../core/shared/element-path-tree'
+import { objectValues } from '../../core/shared/object-utils'
+import { fastForEach } from '../../core/shared/utils'
+import { getConditionalClausePath, ThenOrElse } from '../../core/model/conditionals'
+
+function baseNavigatorDepth(path: ElementPath): number {
+  // The storyboard means that this starts at -1,
+  // so that the scenes are the left most entity.
+  return EP.fullDepth(path) - 1
+}
+
+export function navigatorDepth(
+  navigatorEntry: NavigatorEntry,
+  metadata: ElementInstanceMetadataMap,
+): number {
+  const path = navigatorEntry.elementPath
+  let result: number = baseNavigatorDepth(path)
+  for (const ancestorPath of EP.getAncestors(path)) {
+    const elementMetadata = MetadataUtils.findElementByElementPath(metadata, ancestorPath)
+    if (elementMetadata != null) {
+      // If fragments are not supported, shift the depth back by 1 as they will not be included in the
+      // hierarchy.
+      if (!isFeatureEnabled('Fragment support')) {
+        const isFragment = foldEither(
+          () => false,
+          (e) => isJSXFragment(e),
+          elementMetadata.element,
+        )
+        if (isFragment) {
+          result = result - 1
+        }
+      }
+
+      // A conditional ancestor will shift this by an additional 1, for the clause.
+      if (isFeatureEnabled('Conditional support')) {
+        const isConditional = foldEither(
+          () => false,
+          (e) => isJSXConditionalExpression(e),
+          elementMetadata.element,
+        )
+        if (isConditional) {
+          result = result + 1
+        }
+      }
+    }
+  }
+
+  // For the clause entry itself, this needs to step back by 1.
+  if (
+    isFeatureEnabled('Conditional support') &&
+    isConditionalClauseNavigatorEntry(navigatorEntry)
+  ) {
+    result = result - 1
+  }
+
+  return result
+}
+
+interface GetNavigatorTargetsResults {
+  navigatorTargets: Array<NavigatorEntry>
+  visibleNavigatorTargets: Array<NavigatorEntry>
+}
+
+export function getNavigatorTargets(
+  metadata: ElementInstanceMetadataMap,
+  collapsedViews: Array<ElementPath>,
+  hiddenInNavigator: Array<ElementPath>,
+): GetNavigatorTargetsResults {
+  // Note: This will not necessarily be representative of the structured ordering in
+  // the code that produced these elements.
```

Comment:
Is this still true? I know this comment came from the original source of this code, but I think that was from before we started using a tree here 🤔

Reply:
If it is still true, but only applies to the `const projectTree`, please can you mention that in the comment

Reply:
Addressed in this PR: https://github.com/concrete-utopia/utopia/pull/3399